## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [307]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [308]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [309]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [310]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df["item_id"]

0        EUD29711-63-6U
1        XDA18116-89-4A
2        ABE59463-05-7E
3        ZMM00836-65-0C
4        BZX55559-12-3X
              ...      
54634    UCA28532-31-0B
54635    EBD29277-23-7U
54636    EBD30093-55-9Z
54637    XMZ38239-49-3D
54638    ZCZ05589-50-5D
Name: item_id, Length: 54639, dtype: object

In [311]:
df["subcategory"]

0                  wall art
1                      pens
2        project management
3                      pens
4                      misc
                ...        
54634             logistics
54635      kitchen supplies
54636     bathroom supplies
54637                stands
54638    project management
Name: subcategory, Length: 54639, dtype: object

In [312]:
df["category"]


0              decor
1        consumables
2           software
3        consumables
4        consumables
            ...     
54634       software
54635    consumables
54636    consumables
54637      furniture
54638       software
Name: category, Length: 54639, dtype: object

In [313]:
test1 = df["subcategory"].value_counts().head(15)
test1

subcategory
bathroom supplies     6424
paper products        5874
kitchen supplies      5746
tables                4540
misc                  4286
stands                4201
accounting            3756
project management    3522
plants                3466
wall art              3093
seating               2664
pens                  2603
logistics             1122
telecom                978
audio visual           732
Name: count, dtype: int64

In [314]:
# What three item categories had the most entries?
top_3_cat = df["category"].value_counts().head(3)
top_3_cat


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [315]:
# For the category with the most entries,
# which subcategory had the most entries?
top_cat = top_3_cat.index[0]
top_cat_rows = df[df["category"] == top_cat]
top_sub_4_top_cat = top_cat_rows["subcategory"].value_counts().head(1)
top_sub_4_top_cat


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [316]:
#max_5_entries = df["client_id"].value_counts().head(5)

#max_5_entries


In [317]:
# Which five clients had the most entries in the data?
max_5_entries = df["client_id"].value_counts().head(5)
max_5_entriesdf = max_5_entries.reset_index()
max_5_entriesdf.columns = ['client_id', 'entry_count']
max_5_client_ids = df["client_id"].value_counts().head(5).index
max_5_clients = df[df["client_id"].isin(max_5_client_ids)]
max_5_clients = max_5_clients.merge(max_5_entriesdf, on='client_id', how='right')
best_clients = max_5_clients.drop_duplicates("client_id").sort_values("entry_count", ascending=False)
best_clients


first     last                                           job  \
0      Jessica    Reyes  Production designer, theatre/television/film   
220     Angela  Everett                                     Osteopath   
431      Bryan    Myers                                Chief of Staff   
640  Alexandra    Young                                     Dietitian   
847     Kendra  Garrett                        Pension scheme manager   

                phone                            email  client_id  order_id  \
0          5016915923         jessica.reyes@grimes.net      33615   1327730   
220  729-620-6192x875           aeverett@rodriguez.com      66037   1203693   
431  444.233.9146x233            bmyers3696@wilson.com      46820   7109502   
640  265-666-3155x118           ayoung9158@stevens.com      38378   8539164   
847      808.257.2029  garrettkendra8479@jefferson.com      24741   3699423   

     order_date  order_week  order_year         item_id     category  \
0    2023-05-02          18        2023  BMB23158-88-9A        decor   
220  2023-04-09          14        2023  XUX09951-98-0M  consumables   
431  2023-05-12          19        2023  CXU60946-73-5B  consumables   
640  2023-01-29           4        2023  ZMB00381-41-4E  electronics   
847  2023-05-18          20        2023  ZXX97056-96-8A        decor   

           subcategory  unit_price  unit_cost  unit_weight  qty  line_number  \
0               plants      172.84     154.44         2.79   78            0   
220  bathroom supplies      301.87     166.09         0.26  157            2   
431  bathroom supplies      136.13     111.47         5.65  118            9   
640       audio visual      382.59     313.09         7.97  691            0   
847           wall art        9.17       8.17         0.96   65            5   

     entry_count  
0            220  
220          211  
431          209  
640          207  
847          207

In [318]:
# Store the client ids of those top 5 clients in a list.
best_clients_list = best_clients["client_id"].unique()
best_clients_list


array([33615, 66037, 46820, 38378, 24741], dtype=int64)

In [319]:
# How many total units (the qty column) did the
#client with the most entries order order?
most_entries_client = max_5_clients.iloc[0]
most_entries_client_id = most_entries_client["client_id"]
most_entries_client_id
most_entries_client_total_units = df[df['client_id'] == most_entries_client_id]['qty'].sum()
most_entries_client_total_units

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [320]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df["Subtotal"] = df["qty"] * df["unit_price"]
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   Subtotal  
0                1  115164.00  
1                0     523.95  
2                6     527.28  
3                3    1056.18  
4                1    3902.00  
...            ...        ...  
54634            8    2743.29  
54635            0    9709.73  
54636            6   31188.50  
54637            0     165.76  
54638            5    3103.31  

[54639 rows x 19 columns]

In [321]:
# Assuming df is already defined and contains the required columns
# pound for items 50 pounds or under.

def shipping_price(x):
    if x["unit_weight"] * x["qty"] > 50:
        return (7 * (x["unit_weight"]) * x["qty"])
    else:
        return (10 * x["unit_weight"]) * x["qty"]

df["Shipping Price"] = df.apply(shipping_price, axis=1).round(2)
df["Shipping Price"].sort_values(ascending=True)
#df

2091            0.00
35061           0.00
5874            0.00
3429            0.00
23899           0.00
            ...     
29737    19349241.03
40161    20912865.12
50048    22993284.16
12597    40848526.11
25972    63063389.34
Name: Shipping Price, Length: 54639, dtype: float64

In [322]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
def total_price(x):
    return ((x["Subtotal"] + x["Shipping Price"]) * 0.0925) + x["Subtotal"] + x["Shipping Price"]
df["Total Price"] = df.apply(total_price, axis=1).round(2)
df


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...     category         subcategory  \
0              17        2023  ...        decor            wall art   
1              20        2023  ...  consumables                pens   
2               4        2023  ...     software  project management   
3              17        2023  ...  consumables                pens   
4              21        2023  ...  consumables                misc   
...           ...         ...  ...          ...                 ...   
54634          17        2023  ...     software           logistics   
54635           6        2023  ...  consumables    kitchen supplies   
54636           9        2023  ...  consumables   bathroom supplies   
54637           3        2023  ...    furniture              stands   
54638           1        2023  ...     software  project management   

      unit_price  unit_cost  unit_weight   qty  line_number   Subtotal  \
0        1096.80     762.71         7.50   105            1  115164.00   
1          24.95      15.09         1.49    21            0     523.95   
2          13.52       7.86         1.68    39            6     527.28   
3          36.42      24.85         1.23    29            3    1056.18   
4         195.10     108.17        46.43    20            1    3902.00   
...          ...        ...          ...   ...          ...        ...   
54634      83.13      51.60         2.25    33            8    2743.29   
54635     206.59     175.46        11.70    47            0    9709.73   
54636      65.66      57.31         4.16   475            6   31188.50   
54637       1.48       1.22        18.04   112            0     165.76   
54638       3.01       1.61         2.07  1031            5    3103.31   

       Shipping Price  Total Price  
0             5512.50    131839.08  
1              312.90       914.26  
2              458.64      1077.12  
3              356.70      1543.57  
4             6500.20     11364.40  
...               ...          ...  
54634          519.75      3564.87  
54635         3849.30     14813.24  
54636        13832.00     49184.90  
54637        14143.36     15632.71  


In [323]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
def subtotalandshipping(x):
    return (x["unit_cost"] * x["qty"] + x["Shipping Price"])
df["Total Cost"] = df.apply(subtotalandshipping, axis=1).round(2)

df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number   Subtotal  Shipping Price  Total Price  \
0             7.50   105            1  115164.00         5512.50    131839.08   
1             1.49    21            0     523.95          312.90       914.26   
2             1.68    39            6     527.28          458.64      1077.12   
3             1.23    29            3    1056.18          356.70      1543.57   
4            46.43    20            1    3902.00         6500.20     11364.40   
...            ...   ...          ...        ...             ...          ...   
54634         2.25    33            8    2743.29          519.75      3564.87   
54635        11.70    47            0    9709.73         3849.30     14813.24   
54636         4.16   475            6   31188.50        13832.00     49184.90   
54637        18.04   112            0     165.76        14143.36     15632.71   
54638         2.07  1031            5    3103.31        14939.19     19711.43   

       Total Cost  
0        85597.05  
1          629.79  
2          765.18  
3         1077.35  
4         8663.60  
...           ...  
54634     2222.55  
54635    12095.92  
54636    41054.25  
54637    14280.00  
54638  

In [324]:
# Create a column for the profit of
# each line using line cost and line price
def profitfunc(x):
    return (x["Total Price"] - x["Total Cost"])

df["Profit"] = df.apply(profitfunc, axis=1).round(2)
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   Subtotal  Shipping Price  Total Price  Total Cost  \
0                1  115164.00         5512.50    131839.08    85597.05   
1                0     523.95          312.90       914.26      629.79   
2                6     527.28          458.64      1077.12      765.18   
3                3    1056.18          356.70      1543.57     1077.35   
4                1    3902.00         6500.20     11364.40     8663.60   
...            ...        ...             ...          ...         ...   
54634            8    2743.29          519.75      3564.87     2222.55   
54635            0    9709.73         3849.30     14813.24    12095.92   
54636            6   31188.50        13832.00     49184.90    41054.25   
54637            0     165.76        14143.36     15632.71    14280.00   
54638            5    3103.31        14939.19     19711.43    16599.10   

         Profit  
0      46242.03  
1        284.47  
2        311.94  
3        466.22  
4       2700.80  
...         ...  
54634   1342.32  
54635   2717.32  
54636   8130.65  
54637   1352.71  
54638   3112.33  

[54639 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [325]:

 #Check your work using the totals above
#order ids "2742071", "2173913", "6128929"
#152811.89
#162388.71
#923441.25
totals = {2742071: 152811.89,
          2173913: 162388.71,
          6128929: 923441.25

}
order_totals = df.groupby('order_id')['Total Price'].sum()
for order_id, expected_total in totals.items():
    calculated_total = order_totals.get(order_id, None)
    if calculated_total is not None:
        print(f"Order ID {order_id}: Calculated total is {calculated_total:.2f}, Expected total is {expected_total}. Match: {calculated_total - expected_total < .01}")
    #else:
        #print(f"Order ID {order_id} not found in the DataFrame.")

Order ID 2742071: Calculated total is 152811.89, Expected total is 152811.89. Match: True
Order ID 2173913: Calculated total is 162388.71, Expected total is 162388.71. Match: True
Order ID 6128929: Calculated total is 923441.25, Expected total is 923441.25. Match: True


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [326]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
#best_clients_list
df.set_index(df["client_id"], inplace=True)
total_best_client = df.loc[best_clients_list, "Total Price"].groupby("client_id").sum()
total_best_client

client_id
24741    82268892.02
33615     8377308.52
38378    12906550.87
46820     9743794.36
66037    10259514.79
Name: Total Price, dtype: float64

In [327]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
top_clients_df = df.loc[best_clients_list]
summary_data = top_clients_df.groupby(top_clients_df.index).agg(
    total_units_purchased=pd.NamedAgg(column='qty', aggfunc='sum'),
    total_shipping_price=pd.NamedAgg(column='Shipping Price', aggfunc='sum'),
    total_revenue=pd.NamedAgg(column='Total Price', aggfunc='sum'),
    total_profit=pd.NamedAgg(column='Profit', aggfunc='sum')
)
summary_sorted_by_profit = summary_data.sort_values('total_profit', ascending=False)
summary_sorted_by_profit


total_units_purchased  total_shipping_price  total_revenue  \
client_id                                                               
24741                     239862            5126448.37    82268892.02   
38378                      73667            3429455.40    12906550.87   
66037                      43018            1395151.85    10259514.79   
46820                      75768            1601448.84     9743794.36   
33615                      64313            1828984.89     8377308.52   

           total_profit  
client_id                
24741       36579992.31  
38378        3271829.89  
66037        3255031.81  
46820        2736602.72  
33615        2201994.61

In [332]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
df.rename(columns={
    'qty': 'Total Units Purchased',
    'Shipping Price': 'Total Shipping Price ($M)',
    'Profit': 'Total Profit ($M)'
}, inplace=True)

df['Total Shipping Price ($M)'] = (df['Total Shipping Price ($M)'] / 1_000_000).round(2)
df['Total Profit ($M)'] = (df['Total Profit ($M)'] / 1_000_000).round(2)
df


first       last                      job               phone  \
client_id                                                                      
58515         Donald    Harding             Immunologist  793-904-7725x39308   
37609        Tiffany      Myers          Music therapist    201.442.4543x942   
57113        Shannon     Watson             Immunologist   687.737.9424x8503   
46554         Nathan      Baker    Accounting technician    827-788-8123x012   
92089      Christina   Schwartz             Chiropractor        265-829-3643   
...              ...        ...                      ...                 ...   
29043         Lauren      Reese           Radio producer  977.877.5272x11382   
41908        Derrick      Moore                 Musician        358.661.5483   
35176         Monica  Gutierrez         Graphic designer    294.805.9100x339   
24485          Wanda    Solomon             Toxicologist       (311)767-4924   
24007         Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                         email  client_id  order_id  \
client_id                                                             
58515         harding.donald.7185@sullivan.com      58515   8953482   
37609        myers.t.6537@ferguson-johnson.net      37609   8069089   
57113                    swatson8146@payne.net      57113   1902144   
46554                   bakernathan@benson.com      46554   9031802   
92089      christinaschwartz9252@mcconnell.com      92089   1322274   
...                                        ...        ...       ...   
29043       late_reese_4081@montoya-chavez.org      29043   9021716   
41908             derrick.moore.2602@pope.info      41908   6290153   
35176                gutierrezm3195@morris.org      35176   8692622   
24485                solomonwanda5962@ross.org      24485   7592730   
24007                         belld@galvan.com      24007   7489403   

           order_date  order_week  order_year  ... unit_price unit_cost  \
client_id                                      ...                        
58515      2023-04-28          17        2023  ...    1096.80    762.71   
37609      2023-05-19          20        2023  ...      24.95     15.09   
57113      2023-01-29           4        2023  ...      13.52      7.86   
46554      2023-04-25          17        2023  ...      36.42     24.85   
92089      2023-05-28          21        2023  ...     195.10    108.17   
...               ...         ...         ...  ...        ...       ...   
29043      2023-04-26          17        2023  ...      83.13     51.60   
41908      2023-02-08           6        2023  ...     206.59    175.46   
35176      2023-03-05           9        2023  ...      65.66     57.31   
24485      2023-01-18           3        2023  ...       1.48      1.22   
24007      2023-01-04           1        2023  ...       3.01      1.61   

          unit_weight  Total Units Purchased  line_number   Subtotal  \
client_id                                                              
58515            7.50                    105            1  115164.00   
37609            1.49                     21            0     523.95   
57113            1.68                     39            6     527.28   
46554            1.23                     29            3    1056.18   
92089           46.43                     20            1    3902.00   
...               ...                    ...          ...        ...   
29043            2.25                     33            8    2743.29   
41908           11.70                     47            0    9709.73   
35176            4.16                    475            6   31188.50   
24485           18.04                    112            0     165.76   
24007            2.07                   1031            5    3103.31   

           Total Shipping Price ($M)  Total Price  Total Cost  \
client_id                                                       
58515  

In [ ]:
# Sort the updated data by "Total Profit" form highest to lowest
